vision_transformer/main.py

```diff
@@ -0,0 +1,244 @@
+            contiguous(). \
+            view(batch_size, num_patches, -1)
+
+        return patches
+
+
+class InputEmbedding(nn.Module):
+
+    def __init__(self, args):
+        super(InputEmbedding, self).__init__()
+        self.patch_size = args.patch_size
+        self.n_channels = args.n_channels
+        self.latent_size = args.latent_size
+        use_cuda = not args.no_cuda and torch.cuda.is_available()
+        self.device = torch.device("cuda" if use_cuda else "cpu")
+        self.batch_size = args.batch_size
+        self.input_size = self.patch_size * self.patch_size * self.n_channels
+
+        # Linear projection
+        self.LinearProjection = nn.Linear(self.input_size, self.latent_size)
+        # Class token
+        self.class_token = nn.Parameter(torch.randn(self.batch_size, 1, self.latent_size)).to(self.device)
+        # Positional embedding
+        self.pos_embedding = nn.Parameter(torch.randn(self.batch_size, 1, self.latent_size)).to(self.device)
+
+    def forward(self, input_data):
+        input_data = input_data.to(self.device)
+        # Patchifying the Image
+        patchify = PatchExtractor(patch_size=self.patch_size)
+        patches = patchify(input_data)
+
+        linear_projection = self.LinearProjection(patches).to(self.device)
+        b, n, _ = linear_projection.shape
+        linear_projection = torch.cat((self.class_token, linear_projection), dim=1)
+        pos_embed = self.pos_embedding[:, :n + 1, :]
+        linear_projection += pos_embed
+
+        return linear_projection
+
+
+class EncoderBlock(nn.Module):
+
+    def __init__(self, args):
+        super(EncoderBlock, self).__init__()
+
+        self.latent_size = args.latent_size
+        self.num_heads = args.num_heads
+        use_cuda = not args.no_cuda and torch.cuda.is_available()
+        self.device = torch.device("cuda" if use_cuda else "cpu")
+        self.dropout = args.dropout
+
+        self.norm = nn.LayerNorm(self.latent_size)
+        self.attention = nn.MultiheadAttention(self.latent_size, self.num_heads, dropout=self.dropout)
+        self.enc_MLP = nn.Sequential(
+            nn.Linear(self.latent_size, self.latent_size * 4),
+            nn.GELU(),
+            nn.Dropout(self.dropout),
+            nn.Linear(self.latent_size * 4, self.latent_size),
+            nn.Dropout(self.dropout)
+        )
+
+    def forward(self, emb_patches):
+        first_norm = self.norm(emb_patches)
+        attention_out = self.attention(first_norm, first_norm, first_norm)[0]
+        first_added = attention_out + emb_patches
+        second_norm = self.norm(first_added)
+        mlp_out = self.enc_MLP(second_norm)
+        output = mlp_out + first_added
+
+        return output
+
+
+class ViT(nn.Module):
+    def __init__(self, args):
+        super(ViT, self).__init__()
+
+        self.num_encoders = args.num_encoders
+        self.latent_size = args.latent_size
+        use_cuda = not args.no_cuda and torch.cuda.is_available()
```

Comment:
I don't believe any of the models need to be device aware, you can set this logic from outside the nn.module, same fedback for the embedding and encoderblock